### ISSUE TITLE
Cluster getting killed for unknown reasons in 2.3

### ISSUE THREAD

**Issue by thelinuxlich** [rethinkdb/rethinkdb#5683]
A microservice using a older version of rethinkdbdash tried connecting to my Rethinkdb 2.3 cluster and and made two servers shutdown with this error:

```
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: error: Error in src/arch/runtime/thread_pool.cc at line 367:
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: error: Segmentation fault from reading the address 0x8.
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: error: Backtrace:
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: error: Mon Apr 18 22:42:52 2016
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 1 [0xb1da50]: backtrace_t::backtrace_t() at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 2 [0xb1ddb0]: format_backtrace(bool) at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 3 [0xd69573]: report_fatal_error(char const*, int, char const*, ...) at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 4 [0xa3fce3]: linux_thread_pool_t::fatal_signal_handler(int, siginfo_t*, void*) at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 5 [0x7f78a11f1100]: /lib64/libpthread.so.0(+0xf100) [0x7f78a11f1100] at 0x7f78a11f1100 (/lib64/libpthread.so.0)
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 6 [0xa9e050]: shared_buf_t::size() const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 7 [0x9e7761]: ql::datum_get_array_size(shared_buf_ref_t<char> const&) at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 8 [0x9a0860]: ql::datum_t::obj_size() const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 9 [0x99fa1f]: /usr/bin/rethinkdb() [0x99fa1f] at 0x99fa1f ()
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 10 [0x99fb20]: ql::datum_t::is_ptype() const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 11 [0x9a06d5]: ql::datum_t::is_ptype(std::string const&) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 12 [0x9a61aa]: ql::datum_t::drop_literals_unchecked_stack(bool*) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 13 [0x9a6b3b]: ql::datum_t::drop_literals(bool*) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 14 [0x9a65d6]: ql::datum_t::drop_literals_unchecked_stack(bool*) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 15 [0x9a790e]: ql::datum_t::default_merge_unchecked_stack(ql::datum_t const&) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 16 [0x9a7f2e]: ql::datum_t::merge(ql::datum_t const&) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 17 [0x9b3d9d]: ql::table_t::batched_insert(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> >&&, std::vector<bool, std::allocat
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 18 [0x9588dc]: ql::insert_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 19 [0x8c496a]: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 20 [0x9b1195]: ql::runtime_term_t::eval_on_current_stack(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 21 [0x9b155f]: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 22 [0x800016]: ql::query_cache_t::ref_t::run(ql::env_t*, ql::response_t*) at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 23 [0x80070e]: ql::query_cache_t::ref_t::fill_response(ql::response_t*) at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 24 [0x8c1273]: rdb_query_server_t::run_query(ql::query_params_t*, ql::response_t*, signal_t*) at ??:?
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 25 [0xb19ea7]: void save_exception<void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, 
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 26 [0xb1a450]: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda(
Apr 18 22:42:52 aidax-rethinkdb-1 rethinkdb[4664]: 27 [0xa42108]: coro_t::run() at ??:?
```

**Comment by thelinuxlich** [comment]
Forgot to mention this microservice tried to open a changefeed

**Comment by mike-marcacci** [comment]
This backtrace looks quite similar to the comment at https://github.com/rethinkdb/rethinkdb/issues/5348#issuecomment-186863025

**Comment by thelinuxlich** [comment]
I did a investigation in the microservice and noticed it is using a proxy with the 2.2 version, so it is not related to the js driver, gonna change the issue title

**Comment by thelinuxlich** [comment]
@mlucy actually I updated the proxy and my cluster gets constantly killed so it is not related :( :( :(

Another stacktrace:

```
pr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: error: Error in src/arch/runtime/thread_pool.cc at line 367:
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: error: Segmentation fault from reading the address 0x8.
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: error: Backtrace:
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: error: Tue Apr 19 19:36:33 2016
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 1 [0xb1da50]: backtrace_t::backtrace_t() at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 2 [0xb1ddb0]: format_backtrace(bool) at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 3 [0xd69573]: report_fatal_error(char const*, int, char const*, ...) at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 4 [0xa3fce3]: linux_thread_pool_t::fatal_signal_handler(int, siginfo_t*, void*) at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 5 [0x7f77620ef100]: /lib64/libpthread.so.0(+0xf100) [0x7f77620ef100] at 0x7f77620ef100 (/lib64/libpthread.so.0)
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 6 [0xa9e050]: shared_buf_t::size() const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 7 [0x9e7761]: ql::datum_get_array_size(shared_buf_ref_t<char> const&) at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 8 [0x9a0860]: ql::datum_t::obj_size() const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 9 [0x99fa1f]: /usr/bin/rethinkdb() [0x99fa1f] at 0x99fa1f ()
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 10 [0x99fb20]: ql::datum_t::is_ptype() const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 11 [0x9a06d5]: ql::datum_t::is_ptype(std::string const&) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 12 [0x9a61aa]: ql::datum_t::drop_literals_unchecked_stack(bool*) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 13 [0x9a6b3b]: ql::datum_t::drop_literals(bool*) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 14 [0x9a65d6]: ql::datum_t::drop_literals_unchecked_stack(bool*) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 15 [0x9a790e]: ql::datum_t::default_merge_unchecked_stack(ql::datum_t const&) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 16 [0x9a7f2e]: ql::datum_t::merge(ql::datum_t const&) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 17 [0x9b3d9d]: ql::table_t::batched_insert(ql::env_t*, std::vector<ql::datum_t, std::allocator<ql::datum_t> >&&, std::vector<bool, std::alloca
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 18 [0x9588dc]: ql::insert_term_t::eval_impl(ql::scope_env_t*, ql::args_t*, ql::eval_flags_t) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 19 [0x8c496a]: ql::op_term_t::term_eval(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 20 [0x9b1195]: ql::runtime_term_t::eval_on_current_stack(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 21 [0x9b155f]: ql::runtime_term_t::eval(ql::scope_env_t*, ql::eval_flags_t) const at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 22 [0x800016]: ql::query_cache_t::ref_t::run(ql::env_t*, ql::response_t*) at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 23 [0x80070e]: ql::query_cache_t::ref_t::fill_response(ql::response_t*) at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 24 [0x8c1273]: rdb_query_server_t::run_query(ql::query_params_t*, ql::response_t*, signal_t*) at ??:?
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 25 [0xb19ea7]: void save_exception<void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*,
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 26 [0xb1a450]: void query_server_t::connection_loop<json_protocol_t>(linux_tcp_conn_t*, unsigned long, ql::query_cache_t*, signal_t*)::{lambda
Apr 19 19:36:33 aidax-rethinkdb-2 rethinkdb[11937]: 27 [0xa42108]: coro_t::run() at ??:?
```

**Comment by mlucy** [comment]
Thanks for the bug report @thelinuxlich .  Is the microservice running any insert queries?  If so, would it be possible to post them here?

**Comment by thelinuxlich** [comment]
I have a lot of microservice instances, and some of them are inserting in batches of max 100 records. Unfortunately the stacktrace doesn't help to pinpoint what service is killing the cluster

**Comment by thelinuxlich** [comment]
The most write-heavy microservice inserts in batches, and do something simple like this:

```
r.table("some_table").insert([...],{
    durability: "soft",
    conflict: "update"|"replace"|"error",
    returnChanges: true
})

**Comment by danielmewes** [comment]
I have a theory for where the bug is. Going to build a test binary and send it to @thelinuxlich .

**Comment by danielmewes** [comment]
Handing this over to @nighelles, since it turns out that there are deeper issues with the `changes` return value computation.

I think the issue is with this line https://github.com/rethinkdb/rethinkdb/blob/next/src/rdb_protocol/val.cc#L371 where we move a value out of `pkey_to_change`. If multiple entries in the insert batch have the same primary key and the `update` or `replace` conflict-resolution strategy is used, the value might get moved out multiple times.

**Comment by mlucy** [comment]
@danielmewes -- I just approved the review for this, but @nighelles is out of town.  The fix is in nighelles/5691 if you want to merge it in to do a point release before he's back.

**Comment by danielmewes** [comment]
Fixed in next  7864afb and v2.3.x . Will ship with RethinkDB 2.3.1 this week.

Thanks for reporting the issue @thelinuxlich, and sorry you had to run into it.

**Comment by thelinuxlich** [comment]
Thanks for the super fast fix :)
